eip-getdeployer-draft.md

```diff
@@ -0,0 +1,245 @@
+- Monitoring and forensics: on-chain tooling can flag changes in deployer provenance as an indicator of potentially suspicious redeploys.
+- Lightweight provenance checks: GETDEPLOYER provides a cheaper, semantic lookup complementary to code-hash checks when implementers prefer deployer-level governance.
+
+## Specification
+
+### Overview
+
+To address implementation concerns (stack shape novelty; historical deployer attribution), this proposal introduces a minimal new lookup opcode `GETDEPLOYER`, used in combination with the existing `DELEGATECALL`. The earlier dual-stack-output delegatecall variant has been removed.
+
+### New State Field
+
+Upon contract creation (via `CREATE` or `CREATE2`) at or after the fork activating this EIP, the client MUST persist a new per-account field `deployer` defined as:
+
+- `deployer`: The address in `msg.sender` at the moment the contract is instantiated.
+
+If a contract address is re-created after `SELFDESTRUCT` (possible only with `CREATE2` using the same salt after destruction), the `deployer` field MUST be overwritten with the redeploying contract's (or EOA's) address. Pre-fork contracts MUST return `0x0000000000000000000000000000000000000000` for `deployer` without requiring chain reprocessing.
+
+### Opcode: `GETDEPLOYER`
+
+```text
+GETDEPLOYER (0xXX)    // to be assigned
+Pop:   addr
+Push:  deployer(addr)
+```
+
+Semantics:
+
+1. Pops one stack item `addr`.
+2. If `addr` is not a contract (no code) pushes `0x0`.
+3. If the contract was created before fork activation pushes `0x0`.
+4. Otherwise pushes the stored `deployer` field (left padded to 32 bytes as with addresses elsewhere on the stack).
+
+Gas schedule (subject to tuning during ACD process):
+
+- Cold access: 700 (same order as `EXTCODEHASH`/`EXTCODESIZE`).
+- Warm access: 100 (align with other warm account reads post EIP-2929).
+
+Touch semantics: Reading `deployer` MUST count as an account access for warm/cold classification but MUST NOT mark the account for deletion protection nor alter `accessed_addresses` beyond normal rules.
+
+### Recommended Pattern
+
+Contracts desiring secure delegate execution perform:
+
+```text
+PUSH target          // address of implementation
+DUP1                 // keep target for GETDEPLOYER and later DELEGATECALL
+GETDEPLOYER          // -> deployer
+... (whitelist / verification logic) ...
+// if accepted
+<prepare gas/args>
+DELEGATECALL
+```
+
+This preserves the invariant that `DELEGATECALL` only pushes a single stack item (success) and avoids introducing the first net +2 stack opcode.
+
+
+(Removed) Earlier drafts included a bundled opcode combining delegatecall and deployer lookup. That complexity was dropped; this EIP now scopes strictly to `GETDEPLOYER`.
+
+### CREATE / CREATE2 Hook Pseudocode
+
+```text
+onContractCreation(newAddr, creator):
+   state[newAddr].deployer = creator
+```
+
+### GETDEPLOYER Pseudocode
+
+```text
```

Comment:
generally EL Pseudocode is provided in python in EIP reference, consider changing to that syntax